enhancements/network/cloud-egress-ip.md

```diff
@@ -0,0 +1,652 @@
+means to an end for implementing a OpenShift networking feature in the cloud (in
+this case: egress IP) by the networking stack. As such the primary "user" of
+this cloud component will be the network plugins themselves. This impacts the
+design decision with regards to the cloud component's failure notification, see
+[CRD section](#crd).
+
+##### network plugin <-> cloud API controller
+
+The network plugin needs to remain "source of truth" of the egress IP setup. The
+reason for this is that the network plugin is the one and only component which
+health checks routes to egress nodes and re-balances the egress IP assignment
+depending on the result. This component will need to execute what the network
+plugin has determined is best. This is done since there is no traffic load
+balancing done today for egress IP intended traffic. The network plugin
+constantly initiates a TCP connection to each egress node and depending on the
+success/failure of that: re-balances the egress assignment to other nodes (thus
+avoiding packets being sent to a egress node which is currently un-reachable).
+Going forward, both network plugins will come to support traffic load-balancing
+and a more sophisticated route health detection. That means that the underlying
+networking solution (OVS in the case of both OVN-Kubernetes and openshift-sdn)
+will manage the route health check. It is worth noting that the "amount of
+re-balancing" done by the network plugin is expected to be reduced, since the
+network fabric will be able to load balance traffic between all egress nodes
+provided and always use a healthy node (thus not requiring the network plugin to
+perform an egress IP re-assignment every time a route goes down).
+
+Given the above, it is however important that the network plugin is conveyed the
+cloud's information concerning the nodes' subnet range. This is especially
+evident on GCP, as without it: the network plugin won't be able to perform any
+assignments at all, given the nodes' default `/32` subnet.
+
+#### Inter-component communication
+
+Multiple pieces of information needs to be communicated between the network
+plugins and this cloud component:
+
+1. The cloud's node subnet and capacity information.
+2. The egress IP assignment to be performed by the cloud component
+3. The status of the egress IP assignment performed by the cloud component
+
+Two models have been decided for this: a CRD based approach and a node
+annotated approach.
+
+##### annotated approach
+
+As to satisfy condition 1. the cloud component will annotate all cluster nodes
+with the nodes' cloud subnet as retrieved from the cloud provider's API when it
+is initialized. This annotation will have the following specification:
+
+`cloud.openshift.io/private-ipconfig: {"ifaddr": {"ipv4": "$IPv4_ADDRESS/$IPv4_SUBNET_MASK", "ipv6": "$IPv6_ADDRESS/$IPv6_SUBNET_MASK"}, "capacity": {"ipv4": "$IPv4_CAPACITY",  "ipv6": "$IPv6_CAPACITY"}}`
```

Comment:
need to have `network` in the annotation name since it's our team's annotation.
also something more specific than `private-ipconfig`... I know the data isn't egress-IP-specific, but it's not really that generic either, so it's probably better to give it an egress-IP-specific name.

Reply:
Yeah, sorry I forgot to update this enhancement proposal correctly. The official group name was: `cloud.network.openshift.io`, see: https://github.com/openshift/api/blob/master/cloudnetwork/v1/register.go#L10

I wanted this component to be agnostic of whatever functionality which might come be built on top of what it delivers, hence why I left egress IP out. I can accommodate to that if you feel strongly about it though

Reply:
My point is that it's not _really_ agnostic; the data it contains is exactly the data that the egress IP feature needs, and there are things other users might need that would be API-incompatible with the annotation as you're specifying it here. (Eg, changing the annotation from having only the primary interface's IPs to having a map of interfaces-to-IPs.) So I think it makes more sense to just say that this is for egress IPs.

Reply:
OK, let's use `cloud.network.openshift.io/egress-ipconfig`